l3_dump.py

```diff
@@ -61,76 +219,79 @@ def exec_binary(cmdargs:list):
-# pylint: disable-next=line-too-long
-with subprocess.Popen(["readelf", "-x", ".rodata", PROGRAM_BIN],
-                      stdout=subprocess.PIPE,
-                      stderr=subprocess.PIPE, text=True) as section:
-    stdout, stderr = section.communicate()
-    rodata_offs = int(stdout.split('\n')[2].split()[0], 0)
-
-# pylint: disable-next=line-too-long
-with subprocess.Popen(["readelf", "-p", ".rodata", PROGRAM_BIN],
-                      stdout=subprocess.PIPE,
-                      stderr=subprocess.PIPE, text=True) as p:
-    stdout, stderr = p.communicate()
-
-strings = {}
-for line in stdout.split('\n')[2:]:
-    if line:
-        offs = int(line.split()[1][:-1], 16)
-        s = line.split("]  ")[1]
-        strings[offs] = s
-
-with open(sys.argv[1], 'rb') as file:
-    # Unpack the 1st n-bytes as an L3_LOG{} struct to get a hold
-    # of the fbase-address stashed by the l3_init() call.
-    data = file.read(L3_LOG_HEADER_SZ)
-    idx, loc, fibase, _, _ = struct.unpack('<iiQQQ', data)
-    # print(f"{idx=} {fibase=:x}")
-
-    # pylint: disable=invalid-name
-    nentries = 0
-    loc_prev = 0
-    # Keep reading chunks of log-entries from file ...
-    while True:
-        row = file.read(L3_ENTRY_SZ)
-        len_row = len(row)
-
-        # Deal with eof
-        if not row or len_row == 0 or len_row < L3_ENTRY_SZ:
-            break
-
-        tid, loc, ptr, arg1, arg2 = struct.unpack('<iiQQQ', row)
-
-        # If no entry was logged, ptr to message's string is expected to be NULL
-        if ptr == 0:
-            break
-
-        offs = ptr - fibase - rodata_offs
-
-        # No location-ID will be recorded in log-files if L3_LOC_ENABLED is OFF.
-        if loc == 0:
-            print(f"{tid=} '{strings[offs]}' {arg1=} {arg2=}")
-        elif DECODE_LOC_ID == 0:
-            print(f"{tid=} {loc=} '{strings[offs]}' {arg1=} {arg2=}")
-        elif DECODE_LOC_ID == 1:
-            # ----------------------------------------------------------------
-            # Minor optimization to speed-up unpacking of L3 log-dumps from
-            # tests that log millions of log-entries from the same line-of-code.
-            # If we found a new location-ID, unpack it.
-            if loc != loc_prev:
-                UNPACK_LOC = exec_binary([LOC_DECODER, '--brief', str(loc)])
-
-                # Save-off location-ID and unpacked string for next loop
-                unpack_loc_prev = UNPACK_LOC
-                loc_prev = loc
-            else:
-                UNPACK_LOC = unpack_loc_prev
-            print(f"{tid=} {UNPACK_LOC} '{strings[offs]}' {arg1=} {arg2=}")
-
-        nentries += 1
-
-print(f"Unpacked {nentries=} log-entries.")
+###############################################################################
+# main() driver
+###############################################################################
+# pylint: disable-msg=too-many-locals
+def main():
```

Comment:
Most things are unchanged. Modularized into a function so that this module can be imported by Pytests w/o invoking the older-inline code.